src/erldns_query_throttle.erl

```diff
@@ -11,145 +11,123 @@
-    Res.
-
-is_throttled({127, 0, 0, 1}, ReqCount, _) ->
-    {false, ReqCount + 1};
-is_throttled(Host, ReqCount, LastRequestAt) ->
-    ExceedsLimit = ReqCount >= ?LIMIT,
-    Expired = timestamp() - LastRequestAt > ?EXPIRATION,
-    case Expired of
-        true ->
-            erldns_storage:delete(host_throttle, Host),
-            {false, 1};
+-doc false.
+-spec start_link() -> any().
+start_link() ->
+    case enabled() of
         false ->
-            {ExceedsLimit, ReqCount + 1}
+            ignore;
+        true ->
+            Config = #{
+                scope => erldns,
+                strategy => lru,
+                merger_fun => fun ?MODULE:merger/2,
+                segment_num => ?DEFAULT_BUCKETS,
+                ttl => {seconds, default_ttl() div ?DEFAULT_BUCKETS}
+            },
+            segmented_cache:start_link(?MODULE, Config)
     end.
 
--endif.
-
-%% @doc Sweep the query throttle table for expired host records.
--spec sweep() -> any().
-sweep() ->
-    gen_server:cast(?MODULE, sweep).
-
-%% @doc Stop the query throttle process normally.
--spec stop() -> any().
-stop() ->
-    gen_server:call(?MODULE, stop).
-
-% Gen server hooks
-init([]) ->
-    erldns_storage:create(host_throttle),
-    {ok, Tref} = timer:apply_interval(?SWEEP_INTERVAL, ?MODULE, sweep, []),
-    {ok, #state{tref = Tref}}.
-
-handle_call(stop, _From, State) ->
-    {stop, normal, ok, State}.
-
-handle_cast(sweep, State) ->
-    Keys = erldns_storage:select(host_throttle, [{{'$1', {'_', '$2'}}, [{'<', '$2', timestamp() - ?EXPIRATION}], ['$1']}], infinite),
-    lists:foreach(fun(K) -> erldns_storage:delete(host_throttle, K) end, Keys),
-    {noreply, State}.
-
-handle_info(_Message, State) ->
-    {noreply, State}.
-
-terminate(_Reason, _State) ->
-    erldns_storage:delete_table(host_throttle),
+-doc false.
+-spec merger(non_neg_integer(), non_neg_integer()) -> non_neg_integer().
+merger(_A, B) ->
+    B + 1.
+
+-doc "Throttle the given message if necessary.".
+-spec throttle(dns:message(), Context :: {term(), Host :: host()}) ->
+    ok | throttled.
+throttle(Msg, {udp, Host}) ->
+    case should_throttle(Msg, Host, ?DEFAULT_LIMIT) of
+        {true, ReqCount} ->
+            Metadata = #{protocol => udp, host => Host},
+            telemetry:execute([erldns, pipeline, throttle], #{count => ReqCount}, Metadata),
+            throttled;
+        false ->
+            ok
+    end;
+throttle(_Message, {tcp, _Host}) ->
     ok.
 
-code_change(_PreviousVersion, State, _Extra) ->
-    {ok, State}.
+-spec should_throttle(dns:message(), host(), non_neg_integer()) -> false | {true, non_neg_integer()}.
+should_throttle(Msg, Host, Limit) ->
+    HasAny = lists:any(fun(#dns_query{type = T}) -> T =:= ?DNS_TYPE_ANY end, Msg#dns_message.questions),
+    HasAny andalso should_throttle(Host, Limit).
+
+-spec should_throttle(host(), non_neg_integer()) -> false | {true, non_neg_integer()}.
+should_throttle({127, 0, 0, 1}, _) ->
```

Comment:
If we are doing this for IPv4, we should also do it for IPv6.

Reply:
Pushed 👌🏽